app/api/crud.py

```diff
@@ -66,7 +68,34 @@ async def get(
 
     results = response.json()
 
-    return [
+    # return results
+    results_dicts = [
         {k: v["value"] for k, v in res.items()}
         for res in results["results"]["bindings"]
     ]
+    results_df = pd.DataFrame(results_dicts)
+
+    if results_df.empty:
+        response_obj = [
+            AggDatasetResponse(
+                num_matching_subjects=results_df.squeeze().shape[0]
```

Comment:
> Wouldn't results_df only be empty if there are zero matching subjects?

Yes, I think you're right. I considered just having `num_matching_subjects=0` here, but I figured _if_ for whatever reason an empty list was returned where it wasn't expected, that would be caught by computing the number this way. This may be overcomplicating things though, so we can just hard code it.

> Related: could we just return an empty list in this case? Why does there have to be one (empty) return object?

This is an option. @rmanaem & I were honestly a bit conflicted about what to return when 0 subjects match the query, but we ended up deciding to be consistent with the aggregate response data model since `num_matching_subjects` is available - why wouldn't we take advantage of it in this case? With the current implementation, the response to the user for a well-formed query that doesn't have matching subjects is:
```
{
[
  "dataset": null,
  "dataset_name": null,
  "num_matching_subjects": = 0,
]
}
```
which I think is clearer and more consistent with the format of successful queries than if just `{[]}` was returned.
What do you think @surchs?

Reply:
Ah I see. So right now the API will always return a list of aggDataset objects, correct? That is:

- if I find something, I get a list of (agg)Dataset objects for those datasets in the graph that have any matches
- if I change my criteria, the number of returned dataset objects might change (and also the number of matched subjects in each will change)
- If I find nothing, I will get only one dataset back, but it's not really a dataset, more of a ghost-dataset with no name and zero matching subjects. When I get that, I know that nothing has matched.

I think that last rule is a bit confusing. If you give me an empty ghost-dataset if there are no matches in ANY datasets, why don't you just do this for ALL datasets, i.e. always return a list of all datasets in the graph and only set the num_matches to zero for the ones without matches. To be clear, I don't think that'd better, just more consistent.

To me an empty list would be clear: nothing found. But if you don't like that, you could also make a global response object and nest the dataset objects inside of it. So 

```python
class Response(BaseClass):
  success: bool
  datasets: List[AggDatasetObject]
  # anything else you feel should get passed here
```

Then the outer structure of the response objects stays consistent even when nothing is found and the matched datasets are still in there.

Reply:
> I think that last rule is a bit confusing. If you give me an empty ghost-dataset if there are no matches in ANY datasets, why don't you just do this for ALL datasets, i.e. always return a list of all datasets in the graph and only set the num_matches to zero for the ones without matches.

Okay, this is a good point.

It sounds then like the clearer options would be to either 1) stick with the original no-matches response of an empty list `[]`, or 2) have it return something like:
```
"success": False
"datasets": []
```
where, when there _are_ matching subjects, `success` would evaluate to `True` and `datasets` would be populated with the dataset-specific dicts as normal.

I think probably then, option 1 (empty list) is still simpler - I don't think wrapping the aggregate dataset objects in another response model necessarily adds anything or makes our lives easier at the level of the frontend (to me `"success": True` is implied by a 200 status code anyways).

@rmanaem I know you were also considering the `[]` response initially, do you have any objections to going back to that behaviour? 😛

Reply:
No problem here 😄

Reply:
Sounds good. I can commit those changes :)